File: src/devscope/llm/providers/ollama.py
from __future__ import annotations

import httpx

from devscope.llm.base import LLMRequest, LLMResponse, ProviderError


class OllamaProvider:
    name = "ollama"

    def __init__(
        self,
        *,
        base_url: str = "http://localhost:11434",
        transport: httpx.AsyncBaseTransport | None = None,
        timeout: float = 120.0,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._transport = transport
        self._timeout = timeout

    async def complete(self, req: LLMRequest) -> LLMResponse:
        payload = {
            "model": req.model,
            "messages": [{"role": "user", "content": req.prompt}],
            "stream": False,
            "options": {
                "temperature": req.temperature,
                "num_predict": req.max_tokens,
            },
        }
        async with httpx.AsyncClient(
            base_url=self._base_url,
            transport=self._transport,
            timeout=self._timeout,
        ) as client:
            try:
                resp = await client.post("/api/chat", json=payload)
            except httpx.HTTPError as exc:
                raise ProviderError(f"ollama transport error: {exc}") from exc

        if resp.status_code != 200:
            raise ProviderError(f"ollama {resp.status_code}: {resp.text[:200]}")

        data = resp.json()
        return LLMResponse(
            text=data["message"]["content"],
            prompt_tokens=data.get("prompt_eval_count"),
            output_tokens=data.get("eval_count"),
            cost_usd=0.0,
        )

File: src-tauri/src/llm/ollama.rs
use std::time::Instant;
use serde_json::json;
use async_trait::async_trait;

use crate::error::{AppError, AppResult};
use super::{LlmProvider, LlmRequest, LlmResponse};

pub struct OllamaProvider {
    pub base_url: String,
    pub client: reqwest::Client,
}

impl Default for OllamaProvider {
    fn default() -> Self {
        Self {
            base_url: std::env::var("OLLAMA_BASE_URL")
                .unwrap_or_else(|_| "http://127.0.0.1:11434".to_string()),
            client: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(120))
                .build()
                .expect("reqwest client"),
        }
    }
}

#[async_trait]
impl LlmProvider for OllamaProvider {
    fn name(&self) -> &'static str { "ollama" }

    async fn call(&self, req: &LlmRequest) -> AppResult<LlmResponse> {
        let started = Instant::now();
        let resp = self.client
            .post(format!("{}/api/generate", self.base_url))
            .json(&json!({
                "model": req.model,
                "prompt": req.prompt,
                "stream": false,
            }))
            .send()
            .await
            .map_err(|e| AppError::LlmProvider { provider: "ollama".into(), message: e.to_string() })?;

        if !resp.status().is_success() {
            return Err(AppError::LlmProvider {
                provider: "ollama".into(),
                message: format!("HTTP {}", resp.status()),
            });
        }
        let v: serde_json::Value = resp.json().await?;
        let content = v.get("response").and_then(|x| x.as_str()).unwrap_or("").to_string();
        let duration_ms = started.elapsed().as_millis() as i64;
        Ok(LlmResponse {
            content,
            prompt_tokens: v.get("prompt_eval_count").and_then(|x| x.as_i64()),
            output_tokens: v.get("eval_count").and_then(|x| x.as_i64()),
            cost_usd: Some(0.0),
            duration_ms,
            model: req.model.clone(),
            provider: "ollama".into(),
        })
    }
}
